File: packages/kotlin/cnos-kotlin-firebase/src/main/kotlin/ai/kitsy/cnos/vault/firebase/FirebaseSecretsProvider.kt
package ai.kitsy.cnos.vault.firebase

import ai.kitsy.cnos.SecretVaultProvider
import ai.kitsy.cnos.SecretVaultProviderFactory
import ai.kitsy.cnos.VaultAuthConfig
import ai.kitsy.cnos.VaultDefinition
import ai.kitsy.cnos.vault.gcp.GcpSecretManagerProvider

class FirebaseSecretsProvider(vaultId: String, def: VaultDefinition) : SecretVaultProvider {
    private val gcp = GcpSecretManagerProvider(vaultId, def)

    override fun authenticate(auth: VaultAuthConfig) = gcp.authenticate(auth)
    override fun batchGet(refs: List<String>): Map<String, Any> = gcp.batchGet(refs)
    override fun get(ref: String): Any? = gcp.get(ref)

    companion object {
        fun factory(): SecretVaultProviderFactory = SecretVaultProviderFactory("firebase") { vaultId, def ->
            FirebaseSecretsProvider(vaultId, def)
        }
    }
}

File: packages/php/cnos-firebase/src/FirebaseSecretsProvider.php
<?php

declare(strict_types=1);

namespace Kitsy\Cnos\Firebase;

use Kitsy\Cnos\CnosError;
use Kitsy\Cnos\Gcp\GcpSecretManagerProvider;
use Kitsy\Cnos\SecretVaultProvider;
use Kitsy\Cnos\SecretVaultProviderFactory;
use Kitsy\Cnos\VaultAuthConfig;
use Kitsy\Cnos\VaultDefinition;

/**
 * Firebase secrets are stored in GCP Secret Manager under the Firebase project.
 * This provider delegates to GcpSecretManagerProvider with the Firebase project ID.
 */
class FirebaseSecretsProvider implements SecretVaultProvider
{
    private readonly GcpSecretManagerProvider $delegate;

    public function __construct(
        string $vaultId,
        VaultDefinition $definition,
    ) {
        $cfg        = $definition->auth->config;
        $projectId  = (string) ($cfg['project'] ?? $cfg['projectId'] ?? $cfg['firebaseProjectId'] ?? getenv('FIREBASE_PROJECT_ID') ?: getenv('GOOGLE_CLOUD_PROJECT') ?: '');

        // Build a GCP-compatible definition using the Firebase project ID
        $gcpDef = new VaultDefinition(
            provider: 'gcp-secret-manager',
            auth:     new \Kitsy\Cnos\VaultAuthDefinition(
                method: $definition->auth->method,
                config: array_merge($definition->auth->config, ['project' => $projectId]),
            ),
            mapping:  $definition->mapping,
            fallback: $definition->fallback,
        );

        $this->delegate = new GcpSecretManagerProvider($vaultId, $gcpDef);
    }

    public function authenticate(VaultAuthConfig $auth): void
    {
        $this->delegate->authenticate($auth);
    }

    public function batchGet(array $refs): array
    {
        return $this->delegate->batchGet($refs);
    }

    public function get(string $ref): mixed
    {
        return $this->delegate->get($ref);
    }

    public static function factory(): SecretVaultProviderFactory
    {
        return new SecretVaultProviderFactory(
            provider: 'firebase-secrets',
            create: static fn(string $id, VaultDefinition $def): self => new self($id, $def),
        );
    }
}
